text/0000-repr-type-aliases.md

```diff
@@ -0,0 +1,84 @@
+
+An additional, automatically-applicable lint should be added that warns a user if they use `type = ...` for a well-known primitive type, since adding `type = ` instead of using the type directly introduces the possibility of shadowing. (For example, `#[repr(type = u32)]` becomes `#[repr(u32)]`.)
+
+Similarly, an automatically-applicable lint should be added that warns a user if a `repr` argument references in an-scope type alias without the `type = ` prefix. (For example, `#[repr(MyType)]` becomes `#[repr(type = MyType)]`.)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The requirement for `type =` is unfortunate, but it feels like the best way to ensure that adding new representations isn't a breaking change going forward. Even if we were to decide it weren't a "breaking change," it would still break things anyway, being de-facto breaking.
+
+And, of course, this complicates the compiler. But that's about it.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could always not do this.
+
+But more realistically, here are some alternative designs that were rejected.
+
+## `self::`
+
+We could, instead of using `type =`, require that all types contain a double-colon to indicate they're a path, effectively preventing collisions with arguments that aren't paths. This would require using `self::` for types that are imported in the local scope, and was actually the first proposal of this RFC, but wasn't very well-received.
+
+## Shadowing attributes
+
+Until a future edition, the current set of valid representations could be solidified as taking precedence over any shadowed identifiers. For example, if someone defines `type transparent = u32`, then `repr(transparent)` still means `repr(transparent)` and not `repr(u32)`.
+
+In future editions, we could either:
+
+* Let type aliases shadow all valid representations. This isn't ideal since there is no way to override the shadowing besides nesting your code in a new module and then re-exporting it outside that module, which is very messy.
+* Expand the list of unshadowable representations every edition where necessary.
+
+## Capital letters
+
+You could require that the types start with capital letters- oh, right, `repr(C)` is a thing.
+
+# Prior art
+[prior-art]: #prior-art
+
+No known prior art exists.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None currently.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Future RFCs like [#3607] propose explicit methods of obtaining enum discriminants, and that further justifies the desire to include a change like this. There aren't many other extensions that could be added, however.
```

Comment:
I'm not sure I see a motivating connection between discriminant accessors and aliased discriminant types — they're both great ideas, but feel fairly disjoint to me.

I can imagine that we might like to permit using *any* type that implements `StructuralEq` as a discriminant type. This PR would provide a way to specify such types.

Reply:
I saw this comment and had updated the future possibilities to actually list allowed types, but had totally forgotten to mention `StructuralEq` as well. Will include that in my next pass.